### PR TITLE
Add support of inline noise data augmentation

### PR DESCRIPTION
**Pull request by eu9ene**
OpusTrainer requires whitespace-tokenized alignments to insert inline noise in parallel sentences (Tags modifier).

- Add a new aligner based on [eflomal](https://github.com/robertostling/eflomal/tree/master) library. It is less memory intensive. fast_align failed with OOM for the white-space tokenized corpus
- Use priors from the original corpus to align back-translated corpus. It helps if the back-translated corpus is too small (higher accuracy) and apparently helps with performance if it's too large
- Produce whitespace-tokenized alignments for teacher and student corpus to work with Tags (inline noise) modifier
- Add 3 new alignment steps in Taskcluster (for original, back-translated and student corpus)
- Separate producing shortlist to another step
- Add inline noise augmentation to dataset importer using unsupervised [SimAlign](https://github.com/cisnlp/simalign) lib based on BERT to get alignments there (even though we could use the priors for the small corpus after switching to eflomal SimAlign is supposed to produce higher accuracy which is important for evaluation and also we don't need an extra dependency on alignments step in Taskcluster)
- Adjust augmentation settings
- Add tests for the new alignment tasks

See the paper: SimAlign: [High Quality Word Alignments without Parallel Training Data using Static and Contextualized Embeddings](https://arxiv.org/abs/2004.08728)

Known issues:
- We currently use naive tokenization because it's what OpusTrainer requires. This might produce alignments of lower quality because we don't take into account punctuation and also the vocabulary for eflomal is getting very large. Ideally, we should switch somehow to Moses tokenization that also separates punctuation
- Because we don't do str.split() explicitly there might be some double spaces that might lead to discrepancies in tokenization of alignments and what OpusTrainer does
- We can see some warnings while training `[Trainer] [WARNING] Skipping line because of exception: ValueError('Out-of-bound alignment pairs')` likely related to the different whitespace, but it requires further investigation. There are not too many of them.

I think those issues don't block landing this PR, so we can handle them as follow-up work.

The commits are logically organized, so it should be easier to review.

fixes https://github.com/mozilla/firefox-translations-training/issues/261

### CODE REVIEW

File: pipeline/alignments/generate-shortlist.sh [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+echo "### Subword segmentation with SentencePiece"
+${COMPRESSION_CMD} -dc "${corpus_src}" |
+  parallel --no-notice --pipe -k -j "${threads}" --block 50M "${MARIAN}/spm_encode" --model "${vocab_path}" \
+   >"${dir}/corpus.spm.${SRC}"
+
+${COMPRESSION_CMD} -dc "${corpus_trg}" |
+  parallel --no-notice --pipe -k -j "${threads}" --block 50M "${MARIAN}/spm_encode" --model "${vocab_path}" \
+   >"${dir}/corpus.spm.${TRG}"
+
+python3 align.py \
```
Comment (eu9ene) [inline pipeline/alignments/generate-shortlist.sh:49]: this is the main change in this file. Git did see that it was moved

File: pipeline/train/train.sh [hunk 2]
```diff
@@ -57,15 +58,12 @@ for index in "${!datasets[@]}"; do
...
       train_aln="${alns[index]}"
-      # alignments are currently used only for student that uses one stage of training
       echo "### Generating tsv dataset with alignments ${alignments}"
       paste <(${COMPRESSION_CMD} -dc "${train_set_prefix}.${src}.${ARTIFACT_EXT}") \
             <(${COMPRESSION_CMD} -dc "${train_set_prefix}.${trg}.${ARTIFACT_EXT}") \
             <(${COMPRESSION_CMD} -dc "${train_aln}") \
             >"${tsv_dataset}"
       rm "${train_aln}"
-      # when using tsv, marian requires --guided-alignments argument to be an index of the alignments in the tsv file
-      extra_params+=("--guided-alignment" "2")
```
Comment (gregtatum) [inline pipeline/train/train.sh:68]: Though: I believe this is removed since they are now defined in `pipeline/train/configs/training/student.finetune.yml`

File: tests/test_alignments.py [hunk 3]
```diff
@@ -13,8 +14,97 @@
...
+«Ты злое существо!» - воскликнула Дороти. «Ты не имеешь права забирать у меня туфлю».
+«Я все равно сохраню его, — сказала Ведьма, смеясь над ней, — и когда-нибудь я получу от тебя и другой».
+Это так разозлило Дороти, что она взяла стоявшее рядом ведро с водой и облила им Ведьму, обмочив ее с головы до ног.
+Мгновенно злая женщина громко вскрикнула от страха, а затем, когда Дороти с удивлением посмотрела на нее, Ведьма начала сжиматься и падать.
+«Посмотри, что ты наделал!» она закричала. «Через минуту я растаю».
+«Мне действительно очень жаль», — сказала Дороти, которая была по-настоящему напугана, увидев, что Ведьма тает, как коричневый сахар, у нее на глазах.
+"""
+
+
+def verify_aln(data_dir, dataset, src_corpus, trg_corpus):
```
Comment (gregtatum) [inline tests/test_alignments.py:39]: Suggestion: Please spell out alignment here, it's a bit more to type, but it makes the code easier to read if we avoid acronyms that others may have trouble interpreting.

File: tests/test_data_importer.py [hunk 4]
```diff
@@ -15,12 +15,24 @@
...
+        parts = line.split("\t")
+        src_sent, trg_sent = parts[0], parts[1]
+        min_len = min(len(src_sent.split()), len(trg_sent.split()))
+        aln = " ".join([f"{idx}-{idx}" for idx in range(min_len)])
+        corpus_and_aln.append(f"{line}\t{aln}")
+
+    return corpus_and_aln
+
+
+# it's very slow to download and run BERT on 2000 lines
```
Comment (gregtatum) [inline tests/test_data_importer.py:34]: Praise: Thanks for the mock.

**Review by gregtatum — APPROVED** [review]
Praise: Thanks for taking the time to split this out into logical commits and a few separate PRs. This made the review much easier to do, and I have a lot more confidence over the changes in the code. This all feels maintainable and tested well.

I left a few minor suggestions here and there, but nothing major, mostly around readability and documentation with an eye towards the future horizon of maintainers reading this code.

I'm somewhat unsure of the affects of changing the aligners as a behavior change, so let's make sure and monitor the use of them as we kick up the next training round.

File: pipeline/train/configs/opustrainer/student.yml [hunk 5]
```diff
@@ -8,21 +8,21 @@ train:
...
-#  min_word_length: 2 # Minimum word length for each word in the noisy sentence
-#  max_word_length: 5 # Maximum word length for each word in the noisy sentence
-#  max_words: 6 # Maximum number of words in each noisy sentence
-# TODO: enable inline noise, currently doesn't work because it requires space tokenized alignments
-# TODO: https://github.com/mozilla/firefox-translations-training/issues/261
-#- Tags: 0.05
-#  augment: 1
+- Typos: 0.05
+# inserts new noise sentences
+- Noise: 0.0005
```
[the remark this answers is not in the document]
Comment (gregtatum) [reply pipeline/train/configs/opustrainer/student.yml:16]: I see that you have docs in a follow-up PR, maybe you can refer to those? I would apply the same level of reasoning that I did on the training config yml, a light amount of explanation with a reference to a file with more information. You can apply this feedback in the following PR if you want.

File: tests/test_data_importer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (eu9ene) [reply tests/test_data_importer.py:34]: Not really, it was slow when I ran it locally but I can see that it added just 1 minute on Taskcluster. The model downloading is a lot faster (225mb/s) and inference takes just 1 minute there. https://firefox-ci-tc.services.mozilla.com/tasks/TAOdxKxXTdeczjrJ8eFvtg/runs/1/logs/public/logs/live.log

File: tests/test_data_importer.py [hunk 6]
```diff
@@ -211,6 +238,8 @@ def test_augmentation_mix(data_dir):
 
     run_import("corpus", dataset, prefix)
 
+    AUG_MAX_RATE = 0.35
+    AUG_MIN_RATE = 0.01
     data_dir.print_tree()
```
[the remark this answers is not in the document]
Comment (eu9ene) [reply tests/test_data_importer.py:243]: We already cover the basic scenarios with `run_task` to test integration. These unit tests are lower level and they are supposed to test a variety of test cases. I don't think the coverage would improve if we switched it to the `run_task` here. 

It would be easy to switch but I think we should aim at doing lower-level unit testing on the code level rather than the task level as it's faster, more focused and it's easier to mock things.

File: pipeline/data/dataset_importer.py [hunk 7]
```diff
@@ -28,6 +29,8 @@
 COMP_CMD = os.getenv("COMPRESSION_CMD", "pigz")
 COMP_EXT = os.getenv("ARTIFACT_EXT", "gz")
 
+random.seed(1111)
```
[the remark this answers is not in the document]
Comment (eu9ene) [reply pipeline/data/dataset_importer.py:32]: Interesting question. Just to make it clear: we set random seeds for reproducibility so that if you rerun the training with the same code you get the same results. In general, it is a best practice to set a hardcoded random seed. This code is used for validation/evaluation. For training even if we have the same random seed, the modifiers will be applied to a completely different corpus, so I don't see any issues here. We can totally change it to 2222 or any other number though :)

File: pipeline/data/dataset_importer.py [hunk 8]
```diff
@@ -44,28 +47,35 @@ def __call__(self, batch: List[str]) -> Iterable[str]:
         return batch
 
 
-MIX_PROB = 0.1  # 10% will be augmented in the mix
+MIX_PROB = 0.05  # 5% will be augmented in the mix
+PROB_1 = 1.0  # 100% chance
+PROB_0 = 0.0  # 0% chance
+# probability 1 adds way too much inline noise
```
[the remark this answers is not in the document]
Comment (eu9ene) [reply pipeline/data/dataset_importer.py:53]: it means we don't use prob 1 for inline noise, I can clarify that

File: pipeline/data/dataset_importer.py [hunk 9]
```diff
@@ -90,6 +100,29 @@ def run_cmd(cmd: List[str]):
...
+    from simalign import SentenceAligner
+
+    # We use unsupervised aligner here because statistical tools like fast_align require a large corpus to train on
+    # This is slow without a GPU and is meant to operate only on small evaluation datasets
+
+    # Use BERT with subwords and itermax as it has a higher recall and matches more words than other methods
+    # See more details in the paper: https://arxiv.org/pdf/2004.08728.pdf
+    # and in the source code: https://github.com/cisnlp/simalign/blob/master/simalign/simalign.py
+    # This will download a 700Mb BERT model from Hugging Face and cache it
+    aligner = SentenceAligner(model="bert", token_type="bpe", matching_methods="i")
```
[the remark this answers is not in the document]
Comment (eu9ene) [reply pipeline/data/dataset_importer.py:113]: Yes, it outputs the downloading process to stdout.

File: pipeline/train/train.sh (hunk 2 above)
Comment (eu9ene) [reply pipeline/train/train.sh:68]: yes

File: pipeline/train/configs/opustrainer/student.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (eu9ene) [reply pipeline/train/configs/opustrainer/student.yml:16]: I would prefer consolidating the docs in markdown rather than spreading them through the code. I can add some short notes about the values though, but while we continue tuning them I don't think we'll be able to document a decision behind every hyperparameter we have.

File: pipeline/data/dataset_importer.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (gregtatum) [reply pipeline/data/dataset_importer.py:32]: Replying to this a bit late, but I think doing `random.seed(args.dataset)` would ensure reproducible results, but also ensure that we get good random distributions across different runs, as the dataset names should be stable.

File: pipeline/train/configs/opustrainer/student.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (gregtatum) [reply pipeline/train/configs/opustrainer/student.yml:20]: I'm happy to consolidate on markdown documentation, but would like to make sure we cross link it in the code where it's applicable so it's easy to jump to the docs.
